### PR TITLE
Fix Issue #1809 by determining better type for member references.

### PR DESCRIPTION
**Pull request by wmdietl**
1. Correctly type the "qualifier expression" of a member reference, where "qualifier expression" is the type or expression before `::`.
2. Use separate caches for types for Expressions, Members, and Type Trees.  (Using a single cache masked the bug.)

### CODE REVIEW

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 1]
```diff
@@ -2504,8 +2504,18 @@ protected boolean checkMethodReferenceAsOverride(
...
         // The type of the expression or type use, <expression>::method or <type use>::method.
-        AnnotatedTypeMirror enclosingType =
-                atypeFactory.getAnnotatedType(memberReferenceTree.getQualifierExpression());
+        ExpressionTree qexpTree = memberReferenceTree.getQualifierExpression();
+        AnnotatedTypeMirror enclosingType = atypeFactory.getAnnotatedTypeFromTypeTree(qexpTree);
+        if (qexpTree.getKind() == Tree.Kind.IDENTIFIER) {
+            // This is a hack around the ambiguity between identifiers that could be
+            // types or expressions. Using TreeUtils.isTypeTree unfortunately doesn't
+            // help, as it isn't precise enough.
+            // TODO: determine a nicer way. Is taking the main qualifiers enough?
```
Comment (smillst) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: I think the correct way is to check the `kind` field of the member reference.  I tried this here:
https://github.com/typetools/checker-framework/compare/master...smillst:smillst1809?expand=1  And it seems to work with out changing the ATF caches.

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 2]
```diff
@@ -306,10 +306,25 @@
     private final Map<Tree, AnnotatedTypeMirror> classAndMethodTreeCache;
 
     /**
-     * Mapping from a Tree to its annotated type; before implicits are applied, just what the
-     * programmer wrote.
+     * Mapping from an expression tree to its annotated type; before implicits are applied, just
```
Comment (smillst) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: Why all the changes to caching?

Comment (wmdietl) [reply framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: This issue is only triggered with a small cache. The reason why this worked with a large cache is that getting types for Expressions, Members, and Types used the same cache. An earlier request for a type tree was then found when an expression query happened. With a small cache that earlier value gets evicted.
To prevent this kind of problem completely, this changes it into three separate caches. We should never again run into a situation where something gets cached as an expression and then returned for a TypeTree query.
That would seem like a big win.

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java (hunk 1 above)
Comment (wmdietl) [reply framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: Ah, I only looked in MemberReferenceTree not in JCMemberReference. Thanks, I'll change this.

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java (hunk 2 above)
Comment (smillst) [reply framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: Makes sense.  Thanks for the explanation!

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 3]
```diff
@@ -306,10 +306,25 @@
...
+     * what the programmer wrote. Might alias the same object as fromMemberTreeCache and
```
Comment (smillst) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: The maps might alias each other?  I'm not sure what this last sentence means.

File: framework/src/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java [hunk 4]
```diff
@@ -40,7 +40,7 @@
  *
  * @see org.checkerframework.framework.type.TypeFromTree
  */
-class TypeFromTypeTreeVisitor extends TypeFromTreeVisitor {
+class TypeFromTypeTreeVisitor extends TypeFromExpressionVisitor {
```
Comment (smillst) [inline framework/src/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java]: Why this change?

Comment (wmdietl) [reply framework/src/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java]: This was necessary because `getAnnotatedTypeFromTypeTree` was possibly called on an expression tree.
This is no longer necessary and I undid this.

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java (hunk 3 above)
Comment (wmdietl) [reply framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: The previous situation was that we had one field and one cache.
This changes it to three fields and three caches. This comment was supposed to say that we could also have three fields that all use the same cache.
I think we should just stick with the three separate caches and I'll remove the comment.
